### PR TITLE
18AL: Nashville/Mobile must be terminus if in route

### PR DESCRIPTION
**Pull request by perwestling**
Workaround for 18AL to the problem described in #1096

### CODE REVIEW

**Comment by jenf** [comment]
Looks good to me, waiting for Toby

**Comment by perwestling** [comment]
I think this one and #1093 can be merged (if they are approved). It is trivial to merge them, just put the terminus check of this PR first in revenue_for, then super, and last the handling of 4D.

File: lib/engine/game/g_18_al.rb [hunk 1]
```diff
@@ -19,6 +19,22 @@ class G18AL < Base
...
+        ensure_terminus_first_or_last(route, %w[A4 Q2])
+
+        super
+      end
+
+      private
+
+      def ensure_terminus_first_or_last(route, termini)
+        termini.each do |terminus|
+          index = route.hexes.index { |hex| hex.name == terminus }
```
Comment (jenf) [inline lib/engine/game/g_18_al.rb:33]: Cleaner doing
route.hexes[1...-1].map(&:name).include?(terminus)
or even
(route.hexes[1...-1].map(&:name) || termini).empty?
The above removes the need for the loop

(Note the three dots)

Comment (perwestling) [reply lib/engine/game/g_18_al.rb]: Now rewritten using & operation for arrays.
Yes, it looks better.

File: lib/engine/game/g_18_al.rb [hunk 2]
```diff
@@ -34,6 +34,22 @@ def operating_round(round_num)
...
+        ensure_terminus_first_or_last(route, %w[A4 Q2])
+
+        super
+      end
+
+      private
+
+      def ensure_terminus_first_or_last(route, termini)
+        route.hexes[1...-1].each do |hex|
+          next unless termini.include? hex.name
```
Comment (tobymao) [inline lib/engine/game/g_18_al.rb:48]: @jenf @perwestling parens around this

File: lib/engine/game/g_18_al.rb [hunk 3]
```diff
@@ -34,6 +34,22 @@ def operating_round(round_num)
           [Step::BuyCompany, blocks: true],
         ], round_num: round_num)
       end
+
+      def revenue_for(route)
+        ensure_terminus_first_or_last(route, %w[A4 Q2])
```
Comment (tobymao) [inline lib/engine/game/g_18_al.rb:39]: this doesn't need to be a new method, just put it all in line

File: lib/engine/game/g_18_al.rb (hunk 2 above)
Comment (perwestling) [reply lib/engine/game/g_18_al.rb:48]: Where do you want parentheses?

File: lib/engine/game/g_18_al.rb (hunk 3 above)
Comment (perwestling) [reply lib/engine/game/g_18_al.rb:39]: I can refactor this in the revenue commit.

File: lib/engine/game/g_18_al.rb (hunk 2 above)
Comment (tobymao) [reply lib/engine/game/g_18_al.rb:48]: include?(hex.name)

**Pull request merged**